Fix implementation of String equalsIgnoreCase() and regionMatches()

sharon-wang opened the pull request:
Fixes: #11399

According to the javadoc, `equalsIgnoreCase()` and `regionMatches()` both involve calling `Character.toLowerCase(Character.toUpperCase(character))` on each character in the string. This PR matches that expected behaviour.

Added tests with Turkish characters for `equalsIgnoreCase()`, `regionMatches()` and `compareToIgnoreCase()`.

Uses new helper `charValuesEqualIgnoreCase()` to compare if two characters are equal (case insensitive).

tajila commented:
@JasonFengJ9 can you please review this?

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1815,8 +1815,8 @@ public boolean equalsIgnoreCase(String string) {
 			if (charAtO1Last != charAtO2Last
-					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
-					&& ((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))
+					&& (Character.toLowerCase(toUpperCase(charAtO1Last)) != Character.toLowerCase(toUpperCase(charAtO2Last)))
```
As per https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#equalsIgnoreCase(java.lang.String), 
```
Two characters c1 and c2 are considered the same ignoring case if at least one of the following is true:

The two characters are the same (as compared by the == operator)
Calling Character.toLowerCase(Character.toUpperCase(char)) on each character produces the same result
```
`Character.toLowerCase(Character.toUpperCase(char))` can be used instead of `String. toUpperCase()`.

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1844,9 +1844,10 @@ public boolean equalsIgnoreCase(String string) {
+				if (charAtO1 != charAtO2
+						&& (Character.toLowerCase(toUpperCase(charAtO1)) != Character.toLowerCase(toUpperCase(charAtO2)))
+						&& (charAtO1 <= 255 && charAtO2 <= 255)
+				) {
```
There are two `public boolean equalsIgnoreCase(String string)` within `String.java`, `Java 8` version seems missing.

JasonFengJ9 replied:
Please add the testcase supplied within the issue to `Test_String.java`.

pshipton replied:
There is special handling in the code for the last character. When adding tests, it can test for the turkish i as the only character (which is also the last), and also when something else is the last character.

sharon-wang commented:
Updated to use `Character.toLowerCase(Character.toUpperCase())` in place of `Character.toLowerCase(toUpperCase())` and added test cases to Test_String.java.

It seems we may also want to update `regionMatches()` to call `Character.toLowerCase(Character.toUpperCase())`? As per the API Spec:

http://cr.openjdk.java.net/~iris/se/15/build/latest/api/java.base/java/lang/String.html#regionMatches(boolean,int,java.lang.String,int,int)
> ignoreCase is true and there is some nonnegative integer k less than len such that:
>
>     Character.toLowerCase(Character.toUpperCase(this.charAt(toffset+k))) !=
>     Character.toLowerCase(Character.toUpperCase(other.charAt(ooffset+k)))

sharon-wang commented:
Using a similar test case to #11399, I'm getting

```
"İİİ".regionMatches("İİİ") = true
"İİİ".regionMatches("ııı") = false
"İİİ".regionMatches("Iıı") = false
"İİİ".regionMatches("iıİ") = false

"ııı".regionMatches("İİİ") = false
"ııı".regionMatches("ııı") = true
"ııı".regionMatches("Iıı") = true
"ııı".regionMatches("iıİ") = false

"Iıı".regionMatches("İİİ") = false
"Iıı".regionMatches("ııı") = true
"Iıı".regionMatches("Iıı") = true
"Iıı".regionMatches("iıİ") = false

"iıİ".regionMatches("İİİ") = false
"iıİ".regionMatches("ııı") = false
"iıİ".regionMatches("Iıı") = false
"iıİ".regionMatches("iıİ") = true
```

but the RI returns `true` for all. Will be applying the same fix to `regionMatches()`.

sharon-wang commented:
Updated with `regionMatches()` fix and corresponding tests.

JasonFengJ9 commented on test/functional/Java8andUp/src/org/openj9/test/java/lang/Test_String.java:
```diff
@@ -585,6 +585,8 @@ public void test_equals() {
 	public void test_equalsIgnoreCase() {
 		AssertJUnit.assertTrue("lc version returned unequal to uc", hwlc.equalsIgnoreCase(hwuc));
+		AssertJUnit.assertTrue("Turkish 'ı' lc version returned unequal to 'İ' uc", "\u0131".equalsIgnoreCase("\u0130"));
+		AssertJUnit.assertTrue("Turkish 'i's with 'j' lc version returned unequal to uc", "\u0131\u0049j".equalsIgnoreCase("\u0069\u0049J"));
```
Please add a few more testcases such as `\u0131` & `\u0130` are at middle or end of a string instead of just a single character string.
In addition, could you verify all test cases also pass with `-XX:+CompactStrings`?

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1794,29 +1794,33 @@ public boolean equalsIgnoreCase(String string) {
 
-			if (byteAtO1Last != byteAtO2Last &&
-					toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last))) {
+			if (byteAtO1Last != byteAtO2Last
```
nitpick: please add parentheses for this condition, same for following similar conditions.

sharon-wang commented:
Addressed review comments and ran test cases with and without `-XX:+CompactStrings` -- these changes now show the same behaviour as the RI.

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1794,29 +1794,33 @@ public boolean equalsIgnoreCase(String string) {
-			if (byteAtO1Last != byteAtO2Last &&
-					toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last))) {
+			if ((byteAtO1Last != byteAtO2Last)
+					&& (Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last)))
```
Is this change introduced to fix an existing `equalsIgnoreCase` problem?
`private static int toUpperCase(int codePoint)` invokes `Character.toUpperCase()` internally but avoid it for case `codePoint < 128` apparently for performance reason.

sharon-wang replied:
I tested locally with `toUpperCase(int codePoint)` and `Character.toUpperCase()` and the simple test case results were the same. So it seems it would be better to keep the existing `toUpperCase(int codePoint)` usage, for performance purposes.

This makes me think that we should use `Character.toLowerCase(toUpperCase(int codePoint))` in the spots where we're now using `Character.toLowerCase(Character.toUpperCase())`.

JasonFengJ9 replied:
> This makes me think that we should use Character.toLowerCase(toUpperCase(int codePoint)) in the spots where we're now using Character.toLowerCase(Character.toUpperCase()).

Yup, probably can go `toLowerCase(toUpperCase(int codePoint))`.

JasonFengJ9 replied:
In addition, it seems `String.compareToIgnoreCase()` is doing the right thing, but could you add a test for it against the `Turkish` characters reported as well?

JasonFengJ9 commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1794,29 +1794,33 @@ public boolean equalsIgnoreCase(String string) {
-					&& ((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))
+			if ((charAtO1Last != charAtO2Last)
+					&& (Character.toLowerCase(Character.toUpperCase(charAtO1Last)) != Character.toLowerCase(Character.toUpperCase(charAtO2Last)))
+					&& (charAtO1Last <= 255 && charAtO2Last <= 255)
```
Existing code checks `(charAtO1Last <= 255 && charAtO2Last <= 255)` along with `|| Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))`, now it is required alone as one of conditions, it doesn't seem right.
Does this change pass a personal build test suite like `sanity.functional`?

sharon-wang replied:
Currently waiting on the results of my personal build test suite with the current pushed changes, will update once available.

In the meantime, I tested locally using `(char1 != char2) && (compareValue(char1) != compareValue(char2)` (removed `(charAtO1Last <= 255 && charAtO2Last <= 255)` and swapped `Character.toLowerCase(Character.toUpperCase()` with `compareValue()`), and the expected behaviour was seen. This logic is currently being used here on line `1839`:
https://github.com/eclipse/openj9/blob/f2e8dfa4346808869c9e47822f725a10678ac97d/jcl/src/java.base/share/classes/java/lang/String.java#L1836-L1841

`compareValue(char/int/byte)` internally calls `Character.toLowerCase(Character.toUpperCase(char/int/byte))`

https://github.com/eclipse/openj9/blob/f2e8dfa4346808869c9e47822f725a10678ac97d/jcl/src/java.base/share/classes/java/lang/String.java#L1457-L1478

I'll run another build with these changes to confirm if it's functionally correct.

sharon-wang replied:
Both the build with the currently pushed code and the build with my local changes passed `sanity.functional`. Will be pushing my local changes, which contain the modifications discussed in the review comments.

sharon-wang commented:
Updated PR with fixes. Ran personal build with the `sanity.functional` test suite, which succeeded.

pshipton commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1794,29 +1794,32 @@ public boolean equalsIgnoreCase(String string) {
 
-			if (charAtO1Last != charAtO2Last
-					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
-					&& ((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))
```
@fjeremic fyi. We're losing the <= 256 optimization here, do you think it's worth putting this into compareValue()?

pshipton replied:
I guess that doesn't work as is, we've have to create a new helper function that takes both chars.

fjeremic replied:
Yeah should be fine as the check has moved into `charValuesEqualIgnoreCase`.

sharon-wang commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1477,6 +1477,27 @@ private char compareValue(byte b) {
 		return Character.toLowerCase(Character.toUpperCase(helpers.byteToCharUnsigned(b)));
 	}
 
+	private boolean charValuesEqual(char c1, char c2) {
+		boolean charValuesEqual = false;
+
+		if (c1 == c2) {
+			charValuesEqual = true;
+		} else {
+			// If at least one char is ASCII, converting to upper cases then compare should be sufficient.
+			// If both chars are not in ASCII char set, need to convert to lower case and compare as well.
+			char c1upper = (char) toUpperCase(c1);
+			char c2upper = (char) toUpperCase(c2);
+
+			if (((c1 <= 255 || c2 <= 255) && (c1upper == c2upper))
+					|| (toLowerCase(c1upper) == toLowerCase(c2upper))
+			) {
+				charValuesEqual = true;
+			}
+		}
+
+		return charValuesEqual;
+	}
+
```
@pshipton Here's the new helper. This is now being used anytime we are checking equality for two chars in `equalsIgnoreCase()` and `regionMatches()`.

pshipton replied:
Local testing is fine. I'll be running PR testing anyway.

sharon-wang commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1789,34 +1810,38 @@ public boolean equalsIgnoreCase(String string) {
-			if (byteAtO1Last != byteAtO2Last &&
-					toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last))) {
+			if ((byteAtO1Last != byteAtO2Last)
+					&& (!charValuesEqual(helpers.byteToCharUnsigned(byteAtO1Last), helpers.byteToCharUnsigned(byteAtO2Last)))
```
The testcase from the issue fails with `-XX:+CompactStrings`, but the failure is not occurring in the `enableCompression` code. The failure is still in the "regular" `equalsIgnoreCase()`/`regionMatches()` code when `-XX:+CompactStrings` is used.

So functionally, `charValuesEqual()` is not necessary for `enableCompression` cases (does an extra `toLowerCase()` check) -- I'll revert to the original.

fjeremic commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1477,6 +1477,27 @@ private char compareValue(byte b) {
 		return Character.toLowerCase(Character.toUpperCase(helpers.byteToCharUnsigned(b)));
 	}
 
+	private boolean charValuesEqual(char c1, char c2) {
```
This method should also be `static` as it does not use any fields.

JasonFengJ9 replied:
This method is missing in Java 8 pConfig.

sharon-wang replied:
Added this method to the Java 8 section of String; made `compareValue(int/char/byte)` in Java 9+ section `static` as well, since they do not use any non-static fields (the Java 8 versions are already static).

fjeremic commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -2485,15 +2510,15 @@ public boolean regionMatches(boolean ignoreCase, int thisStart, String string, i
-				if (charAtO1 != charAtO2 &&
-						toUpperCase(charAtO1) != toUpperCase(charAtO2) &&
-						toLowerCase(charAtO1) != toLowerCase(charAtO2)) {
+				if (!charValuesEqualIgnoreCase(charAtO1, charAtO2)) {
```
I think we may be missing the `charAtO1 != charAtO2` check here. I see the check is duplicated in `charValuesEqualIgnoreCase` which seems suboptimal. I think we should take the check out of `charValuesEqualIgnoreCase` and always perform it upfront before calling `charValuesEqualIgnoreCase`, otherwise performance will suck.

sharon-wang replied:
Moved the `char1 != char2` check back outside of `charValuesEqualIgnoreCase()`

fjeremic commented:
FYI tagging @0xdaryl. I forget if `regionMatches` pops up on workloads we track, I believe it does on DT7 IIRC as one of the top five methods. I think this will negatively impact performance. Not sure by how much or that we should even care but just bringing it up for your awareness.

sharon-wang commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -1477,6 +1477,27 @@ private char compareValue(byte b) {
 		return Character.toLowerCase(Character.toUpperCase(helpers.byteToCharUnsigned(b)));
 	}
 
+	private boolean charValuesEqualIgnoreCase(char c1, char c2) {
```
Yes, updated

pshipton commented:
jenkins test sanity plinux jdk8,jdk11

0xdaryl commented:
> I forget if regionMatches pops up on workloads we track, I believe it does on DT7 IIRC as one of the top five methods. I think this will negatively impact performance. Not sure by how much or that we should even care but just bringing it up for your awareness.

Agree.  We won't know its impact until we measure.  It seems prudent to perform some performance runs on workloads OpenJ9 cares about before this is merged.

pshipton commented:
>  It seems prudent to perform some performance runs on workloads OpenJ9 cares about before this is merged.

@sharon-wang pls follow up with Piyush on this.

sharon-wang commented:
LibertyStartupDT7 Results
---
These are the aggregate startup results from 6 iterations of the benchmark.
|   | Mean Startup Time (ms) | Median Startup Time (ms) | Confidence Interval |
|--:|:-:|:-:|:-:|
| **with new changes** | 4653.0 | 4668.5 | 0.0047 |
| **no changes** | 4613.7 | 4599.5 | 0.0056 |
|  **diff** | +39.3 | +69 |
|  **diff %** | +0.85% | +1.50% |

---

Is this regression in startup performance considered slight (acceptable), or will more optimization need to be done?

0xdaryl commented:
@mpirvu: can you provide some guidance on what is considered acceptable for an OpenJ9 startup regression?  In Sharon's results above, it is a little under 1% on average for DT7.

mpirvu commented:
Typically 0.85% is considered noise if you do just 6 iterations of the benchmark, but the quoted confidence interval is suspiciously low. If that holds, then this is a true regression.
To answer @0xdaryl question, the size of the regression is considered against the purpose of the change. If this is a change needed for correctness (or conformance to the spec) and there is no better way of implementing it, we'll have to accept it.
Also, to be on the safe side we need throughput runs as well.

I can do those runs if I am provided with two builds: before and after.

mpirvu commented:
FYI @vijaysun-omr

fjeremic commented:
I dug up some old profiles of DayTrader7. Here are the top methods:

```
Ticks	Pct_Total	%   	Symbol/Functions                                                                                                                                                                                                                                                                                                                                                                                                               	
41298	1.29     	2.09	java/util/concurrent/ConcurrentHashMap.get(Ljava/lang/Object;)Ljava/lang/Object;                                                                                                                                                                                                                                                                                                                                               	
32972	1.03     	1.67	com/ibm/jit/JITHelpers.unsafeObjectShallowCopy64(Ljava/lang/Object;Ljava/lang/Object;J)V                                                                                                                                                                                                                                                                                                                                       	
29315	0.92     	1.48	java/lang/String.equals(Ljava/lang/Object;)Z                                                                                                                                                                                                                                                                                                                                                                                   	
26300	0.82     	1.33	org/eclipse/persistence/internal/descriptors/ObjectBuilder.buildWorkingCopyCloneFromRow(Lorg/eclipse/persistence/queries/ObjectBuildingQuery;Lorg/eclipse/persistence/internal/queries/JoinedAttributeManager;Lorg/eclipse/persistence/internal/sessions/AbstractRecord;Lorg/eclipse/persistence/internal/sessions/UnitOfWorkImpl;Ljava/lang/Object;Lorg/eclipse/persistence/internal/identitymaps/CacheKey;)Ljava/lang/Object;	
22732	0.71     	1.15	com/ibm/tx/jta/impl/TransactionImpl.equals(Ljava/lang/Object;)Z                                                                                                                                                                                                                                                                                                                                                                	
22224	0.70     	1.13	com/ibm/ejs/j2c/SharedPool.getSharedConnection(Ljava/lang/Object;Ljavax/security/auth/Subject;Ljavax/resource/spi/ConnectionRequestInfo;ZLjava/lang/String;II)Lcom/ibm/ws/j2c/MCWrapper;                                                                                                                                                                                                                                       	
18697	0.59     	0.95	org/apache/myfaces/view/facelets/tag/jsf/ComponentTagHandlerDelegate.apply(Ljavax/faces/view/facelets/FaceletContext;Ljavax/faces/component/UIComponent;)V                                                                                                                                                                                                                                                                     	
18324	0.57     	0.93	javax/faces/component/UIComponent.popComponentFromEL(Ljavax/faces/context/FacesContext;)V                                                                                                                                                                                                                                                                                                                                      	
16738	0.52     	0.85	java/util/HashMap.putVal(ILjava/lang/Object;Ljava/lang/Object;ZZ)Ljava/lang/Object;                                                                                                                                                                                                                                                                                                                                            	
16690	0.52     	0.84	org/eclipse/persistence/internal/helper/ConcurrentFixedCache.put(Ljava/lang/Object;Ljava/lang/Object;)V                                                                                                                                                                                                                                                                                                                        	
16057	0.50     	0.81	java/io/Writer.write(Ljava/lang/String;II)V     
```

What I was concerned about was the third one, i.e. `String.equals`. Looking at the code though it does call `regionMatches` but the internal version, i.e.
https://github.com/eclipse/openj9/blob/67bb44416a1caafdbcebdff6a2463b44d82f13cd/jcl/src/java.base/share/classes/java/lang/String.java#L2382-L2413

Going back to the changes being made here, they are in the case sensitive version of this API. This will not be on the critical path of at least DayTrader7. I think we should be ok here.

mpirvu commented:
My own DT7 startup experiments show virtually no difference between the two builds provided by @sharon-wang 
```
Results for JDK=/home/mpirvu/sdks/Sharon/default jvmOpts=-Xmx256m
StartupTime     avg=4021        min=3829        max=4270        stdDev=92.7     maxVar=11.5%    confInt=0.30%   samples=159
Footprint       avg=216777      min=209524      max=225664      stdDev=2423.5   maxVar=7.7%     confInt=0.16%   samples=127
CThreadTime     avg=2304        min=1894        max=3136        stdDev=247.9    maxVar=65.6%    confInt=1.41%   samples=159
ProcessTime     avg=7408        min=6820        max=8660        stdDev=429.7    maxVar=27.0%    confInt=0.79%   samples=146

Results for JDK=/home/mpirvu/sdks/Sharon/withchanges jvmOpts=-Xmx256m
StartupTime     avg=4019        min=3848        max=4293        stdDev=86.5     maxVar=11.6%    confInt=0.28%   samples=160
Footprint       avg=217072      min=210372      max=225436      stdDev=2444.4   maxVar=7.2%     confInt=0.16%   samples=131
CThreadTime     avg=2304        min=1888        max=2993        stdDev=220.8    maxVar=58.5%    confInt=1.25%   samples=159
ProcessTime     avg=7412        min=6890        max=8580        stdDev=375.7    maxVar=24.5%    confInt=0.69%   samples=146

fjeremic commented:
I think we should be ok to proceed here.

sharon-wang commented:
Thank you all for the help with investigating perf! @pshipton Looks like this should be ready to merge.

pshipton commented:
@JasonFengJ9 any further concerns?

@sharon-wang was your last push 2 days ago just a rebase, or did something change?

sharon-wang commented:
> was your last push 2 days ago just a rebase, or did something change?

Yes, it was a rebase to bring sources up to date for the perf tests.

pshipton commented:
In that case I won't bother re-running the PR testing, which passed. It lgtm, once Jason signs off this can be merged. We should also add the change the 0.24 release.

JasonFengJ9 reviewed (approved):
LGTM

Thanks @sharon-wang

sharon-wang commented:
Opened https://github.com/eclipse/openj9/pull/11497 to add this to 0.24.
